### PR TITLE
Enhance etcd lease to add peer url enabled status as an annotation

### LINKED ISSUE #529: [Feature] Enhance etcd lease to add peer url enabled status as an annotation
**Feature (What you would like to be added):**
Add annotation to etcd lease to contain peerUrlTLSEnabled boolean value.

**Motivation (Why is this needed?):**
Currently for a single node etcd cluster, the member peer URL is not TLS enabled. If this cluster is scaled from 1 to 3 then the new members that are started have peer certs configured and already communicate via peer URL that is TLS enabled. Since the first (existing) member's peer URL is not TLS enabled, the new members fail to communicate and enter into `CrashLoopBackOff`.

[etcd-druid/PR#6533](https://github.com/gardener/gardener/pull/6533) addresses this shortcoming. However, it requires that the lease is updated with peerURLTLSEnabled annotation which it will depend upon to determine if corrective action is required (which involves deleting STS).

**Approach/Hint to the implement solution (optional):**

### PR DESCRIPTION
**Pull request by unmarshall**
**What this PR does / why we need it**:
Adds an annotation to etcd lease which indicates if the peer url is TLS enabled.

**Which issue(s) this PR fixes**:
Fixes #529 

**Special notes for your reviewer**:

**Release note**:

```improvement operator
Adds an annotation to etcd lease which indicates if the peer url is TLS enabled.
```

### CODE REVIEW

File: pkg/server/backuprestoreserver.go [hunk 1]
```diff
@@ -210,6 +216,7 @@ func (b *BackupRestoreServer) runServer(ctx context.Context, restoreOpts *brtype
 			}
 			miscellaneous.SleepWithContext(ctx, retryTimeout)
 		}
+		peerUrlTLSEnabled = true
```
Comment (timuthy) [inline pkg/server/backuprestoreserver.go:219]: I think it's not safe to assume that `peerUrlTLS` is enabled when using multi-node is used. It's true for Gardener but not in general.

File: pkg/health/heartbeat/heartbeat.go [hunk 2]
```diff
@@ -129,6 +132,7 @@ func (hb *Heartbeat) RenewMemberLease(ctx context.Context) error {
 	renewedMemberLease.Spec.HolderIdentity = &memberID
 	renewedTime := time.Now()
 	renewedMemberLease.Spec.RenewTime = &metav1.MicroTime{Time: renewedTime}
+	renewedMemberLease.Annotations = hb.metadata
```
Comment (timuthy) [inline pkg/health/heartbeat/heartbeat.go:135]: Please only overwrite the `PeerUrlTLSEnabledKey`, not the entire annotations.

File: pkg/health/heartbeat/heartbeat.go [hunk 3]
```diff
@@ -306,14 +310,17 @@ func DeltaSnapshotCaseLeaseUpdate(ctx context.Context, logger *logrus.Entry, k8s
...
+func RenewMemberLeasePeriodically(ctx context.Context, stopCh chan struct{}, hconfig *brtypes.HealthConfig, logger *logrus.Entry, etcdConfig *brtypes.EtcdConnectionConfig, peerUrlTLSEnabled bool) error {
 
 	clientSet, err := miscellaneous.GetKubernetesClientSetOrError()
 	if err != nil {
 		return fmt.Errorf("failed to create clientset: %v", err)
 	}
 
-	hb, err := NewHeartbeat(logger, etcdConfig, clientSet)
+	metadata := make(map[string]string)
+	metadata[PeerUrlTLSEnabledKey] = strconv.FormatBool(peerUrlTLSEnabled)
```
Comment (timuthy) [inline pkg/health/heartbeat/heartbeat.go:321]:
```suggestion
        metadata := map[string]string{
           PeerUrlTLSEnabledKey: strconv.FormatBool(peerUrlTLSEnabled)
        }
```

**Review by timuthy — CHANGES_REQUESTED** [review]
As far as I can tell, it generally looks good. I added a few comments but haven't tested it though.

**Comment by ishan16696** [comment]
Can you also mention/describe in the PR that why this change is required ?

File: pkg/server/backuprestoreserver.go (hunk 1 above)
Comment (unmarshall) [reply pkg/server/backuprestoreserver.go:219]: Ok you are correct.

File: pkg/health/heartbeat/heartbeat.go (hunk 2 above)
Comment (unmarshall) [reply pkg/health/heartbeat/heartbeat.go]: Fixed now

**Review by timuthy — APPROVED** [review]
/lgtm
/needs second-opinion
@aaronfern can you have a look as well?

File: pkg/server/backuprestoreserver.go (hunk 1 above)
Comment (unmarshall) [reply pkg/server/backuprestoreserver.go]: changed it

**Review by timuthy — APPROVED** [review]
Thank you! 🙂

**Comment by ishan16696** [comment]
Hi @unmarshall ,
Can you please do a force push as pipeline checks are not passing and I'm unable to trigger a new build ?

**Comment by unmarshall** [comment]
> Hi @unmarshall , Can you please do a force push as pipeline checks are not passing and I'm unable to trigger a new build ?

Done

**Comment by unmarshall** [comment]
> While testing I've noticed that while the correct annotation is being set in the single node case, when in the multi node case it's sets the annotation to `false` irrespective of whether TLS is enabled or not
> 
> Can you please relook this?

Ok let me check this once. As you have indicated that reading the config file will be better in this case as this is the single source of truth. I really wish the reading is just done once. Perhaps this could be improved going futher.

**Comment by unmarshall** [comment]
>While testing I've noticed that while the correct annotation is being set in the single node case, when in the multi node case it's sets the annotation to false irrespective of whether TLS is enabled or not

I found the issue. If you trace back the initial point is the command to start the server. So if you see init.go then you will see a method NewBackupRestoreComponentConfig which initializes RestorationConfig by invoking brtypes.NewRestorationConfig  which initializes InitialAdvertisePeerURLs to default value which is http://localhost:2380/
Since we are using this only for multi-node so it gives an incorrect result. I will now correct it and only depend upon the config file.

**Review by aaronfern — APPROVED** [review]
Thanks for the changes @unmarshall!
/lgtm

**Pull request merged**